aten/src/ATen/native/cpu/ActivationKernel.cpp

```diff
@@ -0,0 +1,484 @@
+    } else {
+      max_val_vec = vec256::max(value, max_val_vec);
+    }
+  }
+  scalar_t max_val_arr[Vec::size];
+  max_val_vec.store(max_val_arr);
+  for (int64_t i = 0; i < Vec::size; i++) {
+    if (max_val_arr[i] > max_val) {
+      max_val = max_val_arr[i];
+    }
+  }
+  return max_val;
+}
+
+template <typename scalar_t>
+inline void _vec_mul_scalarsub_write(
+    scalar_t* grad_input_data,
+    scalar_t* grad_data,
+    scalar_t* output_data,
+    int64_t dim_size,
+    scalar_t sum) {
+  using Vec = vec256::Vec256<scalar_t>;
+  Vec sum_vec(sum);
+  int64_t d = 0;
+  for (; d < dim_size; d += Vec::size) {
+    Vec output(0);
+    Vec grad(0);
+    int64_t leftover = _leftover<Vec::size>(d, dim_size);
+    output.load(output_data + d, leftover);
+    grad.load(grad_data + d, leftover);
+    grad = grad - sum_vec;
+    grad = grad * output;
+
+    grad.store(grad_input_data + d, leftover);
+  }
+}
+
+template <typename scalar_t>
+inline void _vec_sub_exp_scalarmul_write(
+    scalar_t* grad_input_data,
+    scalar_t* grad_data,
+    scalar_t* output_data,
+    int64_t dim_size,
+    scalar_t sum) {
+  using Vec = vec256::Vec256<scalar_t>;
+  Vec sum_vec(sum);
+  int64_t d = 0;
+  for (; d < dim_size; d += Vec::size) {
+    Vec output(0);
+    Vec grad(0);
+    int64_t leftover = _leftover<Vec::size>(d, dim_size);
+    output.load(output_data + d, leftover);
+    grad.load(grad_data + d, leftover);
+    output = output.exp();
+    output = output * sum_vec;
+    grad = grad - output;
+
+    grad.store(grad_input_data + d, leftover);
+  }
+}
+
+template <typename scalar_t>
+inline scalar_t _vec_norm_exp_sum_write(
+    scalar_t* output_data,
+    int64_t dim_size,
+    scalar_t* input_data,
+    scalar_t max_input) {
+  using Vec = vec256::Vec256<scalar_t>;
+
+  int64_t d = 0;
+  Vec tmpsum(0);
+  Vec max_input_vec(max_input);
+  scalar_t tmp_sum_arr[Vec::size];
+  scalar_t tmp_sum_scalar = 0;
+  for (; d < dim_size; d += Vec::size) {
+    int64_t leftover = _leftover<Vec::size>(d, dim_size);
+    Vec value(0);
+    value.load(input_data + d, leftover);
+    value = value - max_input_vec;
+    value = value.exp();
+    // Need to do a partial add
+    if (d + Vec::size > dim_size) {
+      scalar_t value_arr[Vec::size];
+      value.store(value_arr);
+      for (int64_t i = 0; i < dim_size - d; i++) {
+        tmp_sum_scalar += value_arr[i];
+      }
+      value.store(output_data + d, leftover);
+    } else {
+      tmpsum = tmpsum + value;
+      value.store(output_data + d);
+    }
+  }
+  tmpsum.store(tmp_sum_arr);
+  for (int64_t i = 0; i < Vec::size; i++) {
+    tmp_sum_scalar += tmp_sum_arr[i];
+  }
+  return tmp_sum_scalar;
+}
+
+template <typename scalar_t>
+inline void _vec_log_softmax_lastdim(
+    scalar_t* input_data_base,
+    scalar_t* output_data_base,
+    int64_t outer_size,
+    int64_t dim_size) {
+  using Vec = vec256::Vec256<scalar_t>;
+  static constexpr int64_t CHUNK_SIZE = (128 / sizeof(scalar_t)) * Vec::size;
```

Comment:
Where's the `128` coming from?

Reply:
Size of a cacheline. 1024bits.

Reply:
A cache line is 64 bytes. But Intel CPUs pull in the adjacent cache lines on a cache miss. So to get good performance when you're not traversing memory sequentially you want to deal with two cache lines at once (128 bytes).